lib/cipherstash/client.rb

```diff
@@ -150,13 +154,7 @@ def create_collection(name, schema)
 
         indexes = schema.fetch("indexes", {}).map do |idx_name, idx_settings|
           {
-            meta: {
-              "$indexId" => SecureRandom.uuid,
-              "$indexName" => idx_name,
-              # TODO: we don't need two keys for filter indexes
-              "$prfKey" => SecureRandom.hex(16),
-              "$prpKey" => SecureRandom.hex(16),
-            },
+            meta: index_meta(idx_settings["kind"], idx_name),
```

Comment:
I think I'm following, but does that mean that we'd need to define the class method on each `Index` subclass when there are really only two variants right now?

What would you think about pulling this into a plain module outside of `Client` instead?

Something like:
```ruby
module CipherStash
  class Index
    module Settings
      def meta(kind, name)
        # ...
      end

      def mapping(kind, schema_settings)
        # ...
      end
    end
  end
end
```

Reply:
Yeah, there's only two variants right now, but I as I said to Fi on her PR earlier, I doubt that bloom filters are the final word in index types.  But the main benefit for defining the class method on all subclasses is that it keeps related information together -- all the things "about" a particular index type are defined in the class that implements that index type, rather than scattered in unrelated places.

For the implementation, I don't recommend copy-pasting the settings generation code into each class.  You could implement private class methods for each variant, (eg `ore_settings` and `filter_settings`) and then just call the relevant method in the `settings` method (or whatever you call it) defined in each index kind subclass.

Reply:
I played around with refactoring this a bit today, but bailed on it since I wasn't super happy with the result.

I tried using the `inherited` callback in the `Index` class (similar to the orm_adapter approach that you linked), but I didn't get it working. The callback never seemed to get executed, so I was most likely missing something.

I did get something working without the `inherited` callback, but it still felt a bit awkward because I only moved `meta` into class methods but not the mapping settings. I didn't move over the mapping settings because I wasn't too sure how to refactor handling `fieldType`. Here's [the code](https://github.com/cipherstash/ruby-client/commit/a8557c030c431f15b2eb8a1e311fe5c940b3de47) in a separate branch if you're curious.

Mind if we pair on this or park it for now and move it into a separate card @mpalmer ?

Reply:
I've gone through that commit you linked, and I can definitely see a couple of places where we could make things a bit tidier.  I think a bit of time pairing will definitely see it sorted.

Reply:
Commit from pairing: https://github.com/cipherstash/ruby-client/pull/13/commits/a762979a75b0183c678441b849dbf7ab92acb0c1